### PR TITLE
scons: fix re-compiles MPCs every time

### LINKED ISSUE #29475: scons: re-compiles MPCs every time
(or retrieves from cache every time)

```python
comma@tici:/data/openpilot$ scons 
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/Makefile' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/main_lat.c' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver_lat.h' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver.pxd' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/lat_model/lat_expl_vde_adj.c' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/lat_model/lat_model.h' from cache
cd /data/openpilot/selfdrive/controls/lib/lateral_mpc_lib && python3 lat_mpc.py
Warning: Please note that the following versions of CasADi  are officially supported: 3.5.6 or 3.5.5 or 3.5.4 or 3.5.3 or 3.5.2 or 3.5.1 or 3.4.5 or 3.4.0.
 If there is an incompatibility with the CasADi generated code, please consider changing your CasADi version.
Version 3.6.3 currently in use.
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/Makefile' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/main_long.c' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver_long.h' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver.pxd' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/long_model/long_expl_vde_adj.c' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/long_model/long_model.h' from cache
cd /data/openpilot/selfdrive/controls/lib/longitudinal_mpc_lib && python3 long_mpc.py
Warning: Please note that the following versions of CasADi  are officially supported: 3.5.6 or 3.5.5 or 3.5.4 or 3.5.3 or 3.5.2 or 3.5.1 or 3.4.5 or 3.4.0.
 If there is an incompatibility with the CasADi generated code, please consider changing your CasADi version.
Version 3.6.3 currently in use.
scons: done building targets.

comma@tici:/data/openpilot$ scons 
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/Makefile' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/main_lat.c' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver_lat.h' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver.pxd' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/lat_model/lat_expl_vde_adj.c' from cache
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/lat_model/lat_model.h' from cache
cd /data/openpilot/selfdrive/controls/lib/lateral_mpc_lib && python3 lat_mpc.py
Warning: Please note that the following versions of CasADi  are officially supported: 3.5.6 or 3.5.5 or 3.5.4 or 3.5.3 or 3.5.2 or 3.5.1 or 3.4.5 or 3.4.0.
 If there is an incompatibility with the CasADi generated code, please consider changing your CasADi version.
Version 3.6.3 currently in use.
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/Makefile' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/main_long.c' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver_long.h' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver.pxd' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/long_model/long_expl_vde_adj.c' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/long_model/long_model.h' from cache
cd /data/openpilot/selfdrive/controls/lib/longitudinal_mpc_lib && python3 long_mpc.py
Warning: Please note that the following versions of CasADi  are officially supported: 3.5.6 or 3.5.5 or 3.5.4 or 3.5.3 or 3.5.2 or 3.5.1 or 3.4.5 or 3.4.0.
 If there is an incompatibility with the CasADi generated code, please consider changing your CasADi version.
Version 3.6.3 currently in use.
scons: done building targets.
```

### PR DESCRIPTION
**Pull request by deanlee**
remove invalid files from generate_files to fix this issue.

resolve https://github.com/commaai/openpilot/issues/29475

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Something still isn't working right. I checked this out in a fully-built openpilot, and had to run twice before everything was built properly:
```
batman:lib$ scons -u -j8 .
scons: Entering directory `/home/batman/twopilot'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/Makefile' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/main_long.c' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver_long.h' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver.pxd' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/long_model/long_expl_vde_adj.c' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/long_model/long_model.h' from cache
cd /home/batman/twopilot/selfdrive/controls/lib/longitudinal_mpc_lib && python3 long_mpc.py
Warning: Please note that the following versions of CasADi  are officially supported: 3.5.6 or 3.5.5 or 3.5.4 or 3.5.3 or 3.5.2 or 3.5.1 or 3.4.5 or 3.4.0.
 If there is an incompatibility with the CasADi generated code, please consider changing your CasADi version.
Version 3.6.3 currently in use.
scons: done building targets.
batman:lib$ scons -u -j8 .
scons: Entering directory `/home/batman/twopilot'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
clang -o selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.o -c -std=gnu11 -DSWAGLOG="\"common/swaglog.h\"" -g -fPIC -O2 -Wunused -Wshadow -Wno-unknown-warning-option -Wno-deprecated-register -Wno-register -Wno-inconsistent-missing-override -Wno-c99-designator -Wno-reorder-init-list -Wno-error=unused-but-set-variable -DSWAGLOG="\"common/swaglog.h\"" -Wno-#warnings -Wno-shadow -Wno-deprecated-declarations -I. -Ithird_party/acados/include -Ithird_party/acados/include/blasfeo/include -Ithird_party/acados/include/hpipm/include -Ithird_party/catch2/include -Ithird_party/libyuv/include -Ithird_party/json11 -Ithird_party/linux/include -Ithird_party/snpe/include -Ithird_party/mapbox-gl-native-qt/include -Ithird_party/qrcode -Ithird_party -Icereal -Iopendbc/can -I/home/batman/.pyenv/versions/3.11.4/include/python3.11 -I/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/numpy/core/include selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c
selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c:8036:28: warning: equality comparison with extraneous parentheses [-Wparentheses-equality]
  if (((__pyx_v_c_mode[0]) == 'f')) {
       ~~~~~~~~~~~~~~~~~~~~^~~~~~
selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c:8036:28: note: remove extraneous parentheses around the comparison to silence this warning
  if (((__pyx_v_c_mode[0]) == 'f')) {
      ~                    ^     ~
selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c:8036:28: note: use '=' to turn this equality comparison into an assignment
  if (((__pyx_v_c_mode[0]) == 'f')) {
                           ^~
                           =
1 warning generated.
scons: done building targets.
batman:lib$ scons -u -j8 .
scons: Entering directory `/home/batman/twopilot'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
scons: `selfdrive/controls/lib' is up to date.
scons: done building targets.
```

**Comment by deanlee** [comment]
it does need to rebuild again because the target of the `env.command` changed.
think this PR should have done its job, the cython warning is another issue.

**Comment by sshane** [comment]
I had to run it three times before it stopped pulling from cache, second were these two:

```
Retrieved `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_ocp_solver_pyx.o' from cache
Retrieved `selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.o' from cache
```

**Comment by sshane** [comment]
worked for me, no cache pulls on the third run

File: selfdrive/controls/lib/lateral_mpc_lib/SConscript [hunk 1]
```diff
@@ -38,9 +38,7 @@ generated_files = [
   f'{gen}/lat_model/lat_expl_vde_adj.c',
 
   f'{gen}/lat_model/lat_model.h',
```
Comment (sshane) [inline selfdrive/controls/lib/lateral_mpc_lib/SConscript:40]:
```suggestion
  f'{gen}/lat_model/lat_model.h',
  f'{gen}/lat_cost/lat_cost.h.h',
```

It looks like they moved these three headers into `lat_cost.h` in the update from https://github.com/commaai/openpilot/pull/28821.

File: selfdrive/controls/lib/longitudinal_mpc_lib/SConscript [hunk 2]
```diff
@@ -44,10 +44,9 @@ generated_files = [
   f'{gen}/long_model/long_expl_vde_adj.c',
 
   f'{gen}/long_model/long_model.h',
-  f'{gen}/long_constraints/long_h_constraint.h',
-  f'{gen}/long_cost/long_cost_y_fun.h',
-  f'{gen}/long_cost/long_cost_y_e_fun.h',
-  f'{gen}/long_cost/long_cost_y_0_fun.h',
+  f'{gen}/long_constraints/long_constraints.h',
+  f'{gen}/long_cost/long_cost.h',
```
Comment (sshane) [inline selfdrive/controls/lib/longitudinal_mpc_lib/SConscript:48]: Didn't check long, but probably also moved to one long_cost.h file

**Review by sshane — CHANGES_REQUESTED** [review]

**Comment by sshane** [comment]
Is there any scons option to throw an error on missing generated files?

**Comment by adeebshihadeh** [comment]
related to that: https://github.com/commaai/openpilot/issues/29500

**Comment by deanlee** [comment]
> Is there any scons option to throw an error on missing generated files?

there is a `TargetNotBuiltWarning`:
`SCons.Warnings.enableWarningClass(SCons.Warnings.TargetNotBuiltWarning)`

https://github.com/commaai/openpilot/issues/29500 enable all warnings, including this one.

File: selfdrive/controls/lib/longitudinal_mpc_lib/SConscript (hunk 2 above)
Comment (sshane) [reply selfdrive/controls/lib/longitudinal_mpc_lib/SConscript:48]: ah I see you just missed the lateral one

**Comment by sshane** [comment]
from a clean state, after a first compile, it still re-compiles these:

```python
batman@workstation-shane:~/openpilot$ scons -j32 
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
clang -o selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_ocp_solver_pyx.o -c -std=gnu11 -DSWAGLOG="\"common/swaglog.h\"" -g -fPIC -O2 -Wunused -Wshadow -Wno-unknown-warning-option -Wno-deprecated-register -Wno-register -Wno-inconsistent-missing-override -Wno-c99-designator -Wno-reorder-init-list -Wno-error=unused-but-set-variable -DSWAGLOG="\"common/swaglog.h\"" -Wno-#warnings -Wno-shadow -Wno-deprecated-declarations -I. -Ithird_party/acados/include -Ithird_party/acados/include/blasfeo/include -Ithird_party/acados/include/hpipm/include -Ithird_party/catch2/include -Ithird_party/libyuv/include -Ithird_party/json11 -Ithird_party/linux/include -Ithird_party/snpe/include -Ithird_party/mapbox-gl-native-qt/include -Ithird_party/qrcode -Ithird_party -Icereal -Iopendbc/can -I/home/batman/.pyenv/versions/3.11.4/include/python3.11 -I/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/numpy/core/include selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c
clang -o selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver_lat.os -c -std=gnu11 -DSWAGLOG="\"common/swaglog.h\"" -DACADOS_WITH_QPOASES -g -fPIC -O2 -Wunused -Werror -Wshadow -Wno-unknown-warning-option -Wno-deprecated-register -Wno-register -Wno-inconsistent-missing-override -Wno-c99-designator -Wno-reorder-init-list -Wno-error=unused-but-set-variable -DSWAGLOG="\"common/swaglog.h\"" -Wno-unused -fPIC -I. -Ithird_party/acados/include -Ithird_party/acados/include/blasfeo/include -Ithird_party/acados/include/hpipm/include -Ithird_party/catch2/include -Ithird_party/libyuv/include -Ithird_party/json11 -Ithird_party/linux/include -Ithird_party/snpe/include -Ithird_party/mapbox-gl-native-qt/include -Ithird_party/qrcode -Ithird_party -Icereal -Iopendbc/can -Ithird_party/json11 selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver_lat.c
clang -o selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver_long.os -c -std=gnu11 -DSWAGLOG="\"common/swaglog.h\"" -DACADOS_WITH_QPOASES -g -fPIC -O2 -Wunused -Werror -Wshadow -Wno-unknown-warning-option -Wno-deprecated-register -Wno-register -Wno-inconsistent-missing-override -Wno-c99-designator -Wno-reorder-init-list -Wno-error=unused-but-set-variable -DSWAGLOG="\"common/swaglog.h\"" -Wno-unused -fPIC -I. -Ithird_party/acados/include -Ithird_party/acados/include/blasfeo/include -Ithird_party/acados/include/hpipm/include -Ithird_party/catch2/include -Ithird_party/libyuv/include -Ithird_party/json11 -Ithird_party/linux/include -Ithird_party/snpe/include -Ithird_party/mapbox-gl-native-qt/include -Ithird_party/qrcode -Ithird_party -Icereal -Iopendbc/can -Ithird_party/json11 selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_solver_long.c
clang -o selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.o -c -std=gnu11 -DSWAGLOG="\"common/swaglog.h\"" -g -fPIC -O2 -Wunused -Wshadow -Wno-unknown-warning-option -Wno-deprecated-register -Wno-register -Wno-inconsistent-missing-override -Wno-c99-designator -Wno-reorder-init-list -Wno-error=unused-but-set-variable -DSWAGLOG="\"common/swaglog.h\"" -Wno-#warnings -Wno-shadow -Wno-deprecated-declarations -I. -Ithird_party/acados/include -Ithird_party/acados/include/blasfeo/include -Ithird_party/acados/include/hpipm/include -Ithird_party/catch2/include -Ithird_party/libyuv/include -Ithird_party/json11 -Ithird_party/linux/include -Ithird_party/snpe/include -Ithird_party/mapbox-gl-native-qt/include -Ithird_party/qrcode -Ithird_party -Icereal -Iopendbc/can -I/home/batman/.pyenv/versions/3.11.4/include/python3.11 -I/home/batman/.pyenv/versions/3.11.4/lib/python3.11/site-packages/numpy/core/include selfdrive/controls/lib/longitudinal_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c
python3 tools/cabana/dbc/generate_dbc_json.py --out tools/cabana/dbc/car_fingerprint_to_dbc.json
Generated and written to tools/cabana/dbc/car_fingerprint_to_dbc.json
scons: done building targets.
batman@workstation-shane:~/openpilot$
```

**Comment by deanlee** [comment]
> from a clean state, after a first compile, it still re-compiles these:
> 

It look like that SCons does not know about the dependency while the SharedLibrary compiled at the first time.  Scons will detect/update new dependencies In the second run, it will be ok on the third run.

run `scons -j20 --debug=explain` at second run:

> scons: rebuilding `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_ocp_solver_pyx.c' because:
>            `third_party/acados/acados_template/acados_ocp_solver_pyx.pyx' changed
>            `third_party/acados/acados_template/acados_solver_common.pxd' changed
>            `selfdrive/controls/lib/lateral_mpc_lib/c_generated_code/acados_solver.pxd' is a new dependency
>            `/home/dean/.cache/pypoetry/virtualenvs/openpilot-j4J0Otje-py3.11/bin/cython' changed 
>  

This doesn't look like a bug, but a feature of scons. I haven't found a good way to solve this yet. I  recommend to merge this PR first. we can come back later to solve the dependency issue of SharedLibrary.

File: selfdrive/controls/lib/lateral_mpc_lib/SConscript [hunk 3]
```diff
@@ -38,9 +38,8 @@ generated_files = [
...
-  f'{gen}/lat_cost/lat_cost_y_fun.h',
-  f'{gen}/lat_cost/lat_cost_y_e_fun.h',
-  f'{gen}/lat_cost/lat_cost_y_0_fun.h',
+  f'{gen}/lat_cost/lat_cost.h',
+  f'{gen}/lat_constraints/lat_constraints.h',
```
Comment (sshane) [inline selfdrive/controls/lib/lateral_mpc_lib/SConscript:42]:
```suggestion
  f'{gen}/lat_constraints/lat_constraints.h',
  f'{gen}/lat_cost/lat_cost.h',
```

File: selfdrive/controls/lib/longitudinal_mpc_lib/SConscript [hunk 4]
```diff
@@ -44,10 +44,9 @@ generated_files = [
...
+
```
Comment (sshane) [inline selfdrive/controls/lib/longitudinal_mpc_lib/SConscript:49]:
```suggestion
  f'{gen}/long_constraints/long_constraints.h',
  f'{gen}/long_cost/long_cost.h',
```

Comment (sshane) [inline selfdrive/controls/lib/longitudinal_mpc_lib/SConscript:49]:
```suggestion
```

**Comment by sshane** [comment]
I agree, we should fix the missing and moved headers first!

**Pull request merged**